### PR TITLE
Exec async

### PR DESCRIPTION
**Pull request by aujjwal-redhat**
Added execute_command_async in the rexe file which will handle the
asynchronous execution of io commands generally.

Fixes: #248

Signed-off-by: Ayush Ujjwal <aujjwal@redhat.com>

### CODE REVIEW

**Comment by aujjwal-redhat** [comment]
Tested the code :
```bash
ret = redant.execute_command_async('python3 /home/async-test/test.py', '10.70.43.115')
print(f'Process id: {ret}')
ret = redant.execute_command_async('python3 /home/async-test/test2.py', '10.70.43.115')
print(f'Process id: {ret}')
ret = redant.execute_command_async('python3 /home/async-test/test3.py', '10.70.43.115')
print(f'Process id: {ret}')
```

**Terminal**
```bash
(redantenv) [root@servera redant]# python3 core/redant_main.py -c ../Utilities/conf.yaml -rf "/home/MyWork/results.txt" -t tests/functional/glusterd/test_sample.py -sp
Process id: 9575

Process id: 9586

Process id: 9597

test_sample-Generic PASS

Table:
 test_sample
 
+-------------+-------------+--------------------+
| Volume Type | Test Result |  Time taken (sec)  |
+-------------+-------------+--------------------+
|   Generic   |     PASS    | 0.9920370578765869 |
+-------------+-------------+--------------------+

Total time taken by the framework is 42.81626772880554
```


```bash
[root@serverb tmp]# ls
redant-pid-9575.txt  systemd-private-9e72f51eaf174e179000fd91fd7a27e9-chronyd.service-66NOih
redant-pid-9586.txt  systemd-private-9e72f51eaf174e179000fd91fd7a27e9-dbus-broker.service-pyii5f
redant-pid-9597.txt  systemd-private-9e72f51eaf174e179000fd91fd7a27e9-systemd-logind.service-MUeYyh
[root@serverb tmp]# ps -aux | grep 'python3'
root        9585  0.4  0.3  10352  7728 ?        S    17:37   0:00 python3 /home/async_test/test.py
root        9596  0.5  0.3  10352  7664 ?        S    17:37   0:00 python3 /home/async_test/test2.py
root        9607  0.4  0.3  10352  7704 ?        S    17:37   0:00 python3 /home/async_test/test3.py
root        9615  0.0  0.0   6180   660 pts/0    S+   17:38   0:00 grep --color=auto python3
[root@serverb tmp]# cat redant-pid-9575.txt 
Test 1 iteration 0
Test 1 iteration 1
Test 1 iteration 2
Test 1 iteration 3
Test 1 iteration 4
Test 1 iteration 5
Test 1 iteration 6
Test 1 iteration 7
Test 1 iteration 8
Test 1 iteration 9
Finished the test in 20 s
[root@serverb tmp]# cat redant-pid-9586.txt 
Test 2 iteration : 0
Test 2 iteration : 1
Test 2 iteration : 2
Test 2 iteration : 3
Test 2 iteration : 4
Test 2 iteration : 5
Test 2 iteration : 6
Test 2 iteration : 7
Test 2 iteration : 8
Test 2 iteration : 9
Test 2 iteration : 10
Test 2 iteration : 11
Test 2 iteration : 12
Test 2 iteration : 13
Test 2 iteration : 14
Finished test 2 in 15 s
[root@serverb tmp]# cat redant-pid-9597.txt 
Test 3 iteration 0
Test 3 iteration 1
Test 3 iteration 2
Test 3 iteration 3
Test 3 iteration 4
Test 3 iteration 5
Test 3 iteration 6
Test 3 iteration 7
Test 3 iteration 8
Test 3 iteration 9
Test 3 iteration 10
Test 3 iteration 11
Test 3 iteration 12
Test 3 iteration 13
Test 3 iteration 14
Test 3 iteration 15
Test 3 iteration 16
Test 3 iteration 17
Test 3 iteration 18
Test 3 iteration 19
Test 3 iteration 20
Test 3 iteration 21
Test 3 iteration 22
Test 3 iteration 23
Test 3 iteration 24
Test 3 finished in 25 s
```

**Comment by aujjwal-redhat** [comment]
@srijan-sivakumar One thing that I have noted is the process running in the background has a pid(returned) + 10.

For ex:
```bash
(testvenv) [root@servera test]# python3 test.py 
Running python3 /home/async_test/test.py
10049

Running python3 /home/async_test/test2.py
10060

Running python3 /home/async_test/test3.py
10071

Running cat /home/test.txt
10082
```

```
[root@serverb async_test]# ps -aux | grep python3
root       10059  2.0  0.3  10352  7668 ?        S    18:00   0:00 python3 /home/async_test/test.py
root       10070  1.0  0.3  10352  7700 ?        S    18:00   0:00 python3 /home/async_test/test2.py
root       10081  1.0  0.4  10352  7844 ?        S    18:00   0:00 python3 /home/async_test/test3.py
root       10095  0.0  0.0   6180   724 pts/2    S+   18:00   0:00 grep --color=auto python3
```

**Comment by aujjwal-redhat** [comment]
Will check if it is fine or not.

**Comment by aujjwal-redhat** [comment]
Apparently, the background processes run on different pids and `$$` returns the pid of the current shell. More info can be referred in this [doc](https://unix.stackexchange.com/questions/194081/how-can-a-background-process-know-its-own-pid/194083#:~:text=Normally%2C%20you%20can%20get%20your,script%20when%20%24%24%20is%20used.&text=But%20each%20of%20those%20background%20processes%20gets%20the%20value%205253%20with%20%24%24.)

This should not hurt the execution as it still running in the background and returning the pid of the current shell and storing that with the same name as well.

**Review by srijan-sivakumar — CHANGES_REQUESTED** [review]

File: common/rexe.py [hunk 1]
```diff
@@ -131,6 +131,63 @@ def execute_command(self, cmd, node):
...
+                    pkey=mykey,
+                )
+                self.logger.debug(f"SSH connection to {node} is successful.")
+                self.node_dict[node] = node_ssh_client
+            except Exception as e:
+                self.logger.error(f"Connection failure. Exceptions {e}.")
+            # On rebooting the node
+            _, stdout, stderr = self.node_dict[node].exec_command(cmd)
+            pid = stdout.readline()
+            self.logger.info(f"Process ID for the {cmd} : {pid}")
```
[the remark this answers is not in the document]
Comment (aujjwal-redhat) [reply common/rexe.py:186]: Ack

**Comment by srijan-sivakumar** [comment]
> Apparently, the background processes run on different pids and `$$` returns the pid of the current shell. More info can be referred in this [doc](https://unix.stackexchange.com/questions/194081/how-can-a-background-process-know-its-own-pid/194083#:~:text=Normally%2C%20you%20can%20get%20your,script%20when%20%24%24%20is%20used.&text=But%20each%20of%20those%20background%20processes%20gets%20the%20value%205253%20with%20%24%24.)
> 
> This should not hurt the execution as it still running in the background and returning the pid of the current shell and storing that with the same name as well.

That might not be a good way to do that..how about we create a subshell and then go ahead and execute the command, so the command would be of the form,  `sh -c 'echo $$; exec {cmd} > /tmp/redant-pid-$$.txt 2>&1 &'`

**Comment by srijan-sivakumar** [comment]
@aujjwal-redhat , after much deliberation and checking how to achieve the result in case of a shell environment, it feels like too much to get the desired result. ( Too convoluted ). Hence after checking the workings of Paramiko and going through its libraries, it is noted that we can get the same async behavior with certain modifications of the existing functions already under paramiko. The new PR is under : https://github.com/srijan-sivakumar/redant/pull/252. I'm closing this PR. Though a good approach but we needed something clean, I guess we should have started with Paramiko first before going over to the shell changes.

**Pull request closed without merging**